pkg/tsdb/prometheus/resource/resource.go

```diff
@@ -55,113 +49,49 @@ func delStopHeaders(header http.Header) {
 	}
 }
 
-func addHeaders(header http.Header, toAdd map[string]string) {
-	for k, v := range toAdd {
-		header.Add(k, v)
-	}
-}
-
-func normalizeReqHeaders(headers map[string][]string) map[string]string {
-	h := make(map[string]string, len(headers))
-	for k, v := range headers {
-		h[k] = strings.Join(v, ",")
-	}
-	return h
-}
-
 func New(
-	httpClientProvider httpclient.Provider,
-	cfg *setting.Cfg,
-	features featuremgmt.FeatureToggles,
+	httpClient *http.Client,
 	settings backend.DataSourceInstanceSettings,
 	plog log.Logger,
 ) (*Resource, error) {
-	var jsonData map[string]interface{}
-	if err := json.Unmarshal(settings.JSONData, &jsonData); err != nil {
-		return nil, fmt.Errorf("error reading settings: %w", err)
-	}
-
-	customHeaders := make(map[string]string)
-	var jsonDataMap map[string]interface{}
-
-	err := json.Unmarshal(settings.JSONData, &jsonDataMap)
+	jsonData, err := utils.GetJsonData(settings)
 	if err != nil {
 		return nil, err
 	}
-
-	index := 1
-	for {
-		headerNameSuffix := fmt.Sprintf("httpHeaderName%d", index)
-		headerValueSuffix := fmt.Sprintf("httpHeaderValue%d", index)
-
-		key := jsonDataMap[headerNameSuffix]
-		if key == nil {
-			// No (more) header values are available
-			break
-		}
-
-		if val, ok := settings.DecryptedSecureJSONData[headerValueSuffix]; ok {
-			switch k := key.(type) {
-			case string:
-				customHeaders[k] = val
-			}
-		}
-		index++
-	}
+	httpMethod, _ := maputil.GetStringOptional(jsonData, "httpMethod")
 
 	return &Resource{
-		log:           plog,
-		provider:      client.NewProvider(settings, jsonData, httpClientProvider, cfg, features, plog),
-		customHeaders: customHeaders,
+		log:        plog,
+		promClient: client.NewClient(httpClient, httpMethod, settings.URL),
 	}, nil
 }
 
-func (r *Resource) Execute(ctx context.Context, req *backend.CallResourceRequest) (int, []byte, error) {
+func (r *Resource) Execute(ctx context.Context, req *backend.CallResourceRequest) (*backend.CallResourceResponse, error) {
 	delHopHeaders(req.Headers)
 	delStopHeaders(req.Headers)
-	addHeaders(req.Headers, r.customHeaders)
-	client, err := r.provider.GetClient(normalizeReqHeaders(req.Headers))
-	if err != nil {
-		return 500, nil, err
-	}
 
-	return r.fetch(ctx, client, req)
-}
-
-func (r *Resource) fetch(ctx context.Context, client *client.Client, req *backend.CallResourceRequest) (int, []byte, error) {
 	r.log.Debug("Sending resource query", "URL", req.URL)
-	resp, err := client.QueryResource(ctx, req)
+	resp, err := r.promClient.QueryResource(ctx, req)
 	if err != nil {
-		statusCode := 500
-		if resp != nil {
-			statusCode = resp.StatusCode
-		}
-		return statusCode, nil, err
+		return nil, fmt.Errorf("error querying resource: %v", err)
 	}
 
 	defer func() {
-		err = resp.Body.Close()
-	}()
-
-	// Check that the server actually sent compressed data
-	var reader io.ReadCloser
-	switch resp.Header.Get("Content-Encoding") {
```

Comment:
This does not seem to be needed. If the server responds with gzipped data, we don't care here as we don't really need to read it into specific representation. I assume problem before was that we did not sent the response headers back which would probably mean it wouldn't be decoded in the browser.

@marefr I assume this is correct thing to do. I see in https://github.com/grafana/grafana/blob/57fcfd578dd43fae2b2df335fec7536fef7dd1e3/pkg/middleware/gziper.go#L44 that we don't gzip these paths but my assumption is so that we don't accidentally double gzip and that it is ok to just transparently send the response donwstream. On the other hand it seems like it would make more sense to check the response headers if the response is already gziped and decide based on that?

Reply:
Hmm. Good question. You probably want/need to test that to verify it works as expected. Given gzip header are returned and the bytes is gzipped it should work.